lightning/src/routing/router.rs

```diff
@@ -30,47 +30,60 @@ use crate::prelude::*;
 use crate::sync::Mutex;
 use alloc::collections::BinaryHeap;
 use core::cmp;
-use core::ops::Deref;
+use core::ops::{Deref, DerefMut};
 
 /// A [`Router`] implemented using [`find_route`].
 pub struct DefaultRouter<G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref> where
 	L::Target: Logger,
 	S::Target: for <'a> LockableScore<'a>,
+	S::Target: Score,
 {
 	network_graph: G,
 	logger: L,
 	random_seed_bytes: Mutex<[u8; 32]>,
-	scorer: S
+	scorer: S,
+	score_params: <S::Target as Score>::ScoreParams
 }
 
 impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref> DefaultRouter<G, L, S> where
 	L::Target: Logger,
 	S::Target: for <'a> LockableScore<'a>,
+	S::Target: Score,
 {
 	/// Creates a new router.
-	pub fn new(network_graph: G, logger: L, random_seed_bytes: [u8; 32], scorer: S) -> Self {
+	pub fn new(
+		network_graph: G,
+		logger: L,
+		random_seed_bytes: [u8; 32],
+		scorer: S,
+		score_params: <S::Target as Score>::ScoreParams
+	) -> Self {
 		let random_seed_bytes = Mutex::new(random_seed_bytes);
-		Self { network_graph, logger, random_seed_bytes, scorer }
+		Self { network_graph, logger, random_seed_bytes, scorer, score_params }
 	}
 }
 
-impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref> Router for DefaultRouter<G, L, S> where
+impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref> Router for DefaultRouter< G, L, S> where
 	L::Target: Logger,
 	S::Target: for <'a> LockableScore<'a>,
+	S::Target: Score,
 {
 	fn find_route(
-		&self, payer: &PublicKey, params: &RouteParameters, first_hops: Option<&[&ChannelDetails]>,
+		&self,
+		payer: &PublicKey,
+		params: &RouteParameters,
+		first_hops: Option<&[&ChannelDetails]>,
 		inflight_htlcs: &InFlightHtlcs
 	) -> Result<Route, LightningError> {
 		let random_seed_bytes = {
 			let mut locked_random_seed_bytes = self.random_seed_bytes.lock().unwrap();
 			*locked_random_seed_bytes = Sha256::hash(&*locked_random_seed_bytes).into_inner();
 			*locked_random_seed_bytes
 		};
-
 		find_route(
 			payer, params, &self.network_graph, first_hops, &*self.logger,
 			&ScorerAccountingForInFlightHtlcs::new(self.scorer.lock(), inflight_htlcs),
+			&self.score_params,
```

Comment:
this change gives the error of 
<img width="1027" alt="image" src="https://user-images.githubusercontent.com/906931/235962423-cd35a8f7-8cfa-4db1-8a0e-af26dceba2fa.png">

Reply:
the type here should already be correct, but the issue is that the compiler doesnt think that it can coerce `Score` into `LockableScore<'_>::Locked`.